File: practice_java_program/notpad/LongestSubstringwithoutreturnHashMap.java
import java.util.*;

public class LongestSubstringwithoutreturnHashMap{
    public static void main(String[] args) {

        String s = "abcabcbb";

        HashMap<Character, Integer> map = new HashMap<>();

        int left = 0;
        int maxLength = 0;
        int startIndex = 0;

        for (int right = 0; right < s.length(); right++) {

            char ch = s.charAt(right);

            if (map.containsKey(ch)) {
                left = Math.max(left, map.get(ch) + 1);
            }

            map.put(ch, right);

            if (right - left + 1 > maxLength) {
                maxLength = right - left + 1;
                startIndex = left;
            }
        }

        String result = s.substring(startIndex, startIndex + maxLength);

        System.out.println("Longest Substring = " + result);
        System.out.println("Length = " + maxLength);
    }
}

File: python_oriented_interview/notpad/LongestSubstringwithoutreturnHashMap.py
s = "abcabcbb"
map_dict = {}
left = 0
maxLength = 0
startIndex = 0

for right in range(len(s)):
    ch = s[right]
    if ch in map_dict:
        left = max(left, map_dict[ch] + 1)

    map_dict[ch] = right

    if right - left + 1 > maxLength:
        maxLength = right - left + 1
        startIndex = left

result = s[startIndex : startIndex + maxLength]
print("Longest Substring =", result)
print("Length =", maxLength)
